![Ironhack logo](https://i.imgur.com/1QgrNNw.png)

<body>
    <p style="font-size:28px;text-align:center"><b>Project 02 | Data Cleaning & Manipulation</b></p>
</body>

# Introduction

The objective of this projects was to answer a problem statement, practicing data cleaning and manipulation.

---

<body>
    <p style="font-size:20px"><b>Problem Statement</b></p>
</body>

_What are the most common characteristics of people involved in shark incidents in the history?_

---

To answer this problem, the following characteristics will be analyzed:
- Gender
- Age
- Activity

# Setup

## Import

In [1]:
import pandas as pd
import numpy as np
import re

## Load the dataset

In [2]:
# Load the dataset from a Excel file into a Pandas DataFrame
df = pd.read_excel('GSAF5.xls')

In [3]:
# Check information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25792 entries, 0 to 25791
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8771 non-null   object 
 1   Date                    6558 non-null   object 
 2   Year                    6556 non-null   float64
 3   Type                    6552 non-null   object 
 4   Country                 6508 non-null   object 
 5   Area                    6091 non-null   object 
 6   Location                6008 non-null   object 
 7   Activity                6002 non-null   object 
 8   Name                    6343 non-null   object 
 9   Sex                     5987 non-null   object 
 10  Age                     3660 non-null   object 
 11  Injury                  6528 non-null   object 
 12  Fatal (Y/N)             6006 non-null   object 
 13  Time                    3139 non-null   object 
 14  Species                 3610 non-null 

### Conclusions about the dataset for cleaning purposes
1. Seeing the information about the dataset, it is clear that there are a lot of rows that only contains missing values `NaN`.
2. The columns 'Unnamed: 22' and 'Unnamed: 23' probably do not have any information since they have 1 and 2 non-null values, respectively.

## Create a backup version of the raw dataset

In [4]:
# Create a backup version
df_bkp = df.copy()

# General Data Cleaning

## Headers

In [5]:
# Remove unnecessary spaces, put everything in lower case and replace spaces with underscores
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

In [6]:
# Rename the column 'fatal_(y/n)' to make it simpler
df = df.rename(columns={'fatal_(y/n)': 'fatal'})
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal', 'time', 'species',
       'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

## Columns that are not relevant to the analysis

As mentioned above, the columns `unnamed:_22` and `unnamed:_23` probably do not have any relevant information, so they can be removed.

In [7]:
# Remove the columns 'unnamed:_22' and 'unnamed:_23'
df = df.drop(columns=['unnamed:_22', 'unnamed:_23'])

# Check the result
df.head()

case_number         date    year        type    country             area  \
0    2020.08.20  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1    2020.08.14  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2    2020.08.10  10-Aug-2020  2020.0    Provoked        USA          Florida   
3    2020.08.02  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  2020.07.31.c  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity             name  \
0         New Smyrna Beach, Volusia County  Boogie boarding   Carolina Jones   
1             Shelly Beach, Port Macquarie          Surfing  Chantelle Doyle   
2  Off Gasparilla Island, Charlotte County          Fishing             male   
3                   Candle Reef, St. Croix       Snorkeling     Melony Klein   
4         New Smyrna Beach, Volusia County          Surfing      Megan Tossi   

  sex  ... fatal   time               species  \
0   F  ...     N  11h00                   NaN   
1   F  ...     N  09h30  White shark, 2-to 3m   
2   M  ...     N  16h00    Blacktip shark, 6'   
3   F  ...     N  14h00       Nurse shark, 5'   
4   F  ...     N  17h00                   NaN   

            investigator_or_source                         pdf  \
0  K. McMurray, TrackingSharks.com          2020.08.20-NSB.pdf   
1                   B. Myatt, GSAF  2020.08.14-ShellyBeach.pdf   
2  K. McMurray, TrackingSharks.com     2020.08.10-Provoked.pdf   
3  K. McMurray, TrackingSharks.com        2020.08.02-Klein.pdf   
4  K. McMurray, TrackingSharks.com     2020.07.31.c-Tossi..pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.20   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.14   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.10   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.02   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  2020.07.31.c   

  case_number.2 original_order  
0    2020.08.20         6559.0  
1    2020.08.14         6558.0  
2    2020.08.10         6557.0  
3    2020.08.02         6556.0  
4  2020.07.31.c         6555.0  

[5 rows x 22 columns]

Checking the table above, it is clear that the columns `investigator_or_source`, `pdf`, `href_formula` and `href` will not be relevant to the analysis since they are just information about the source of each incident. Therefore, these columns can be removed.

In [8]:
# Remove the columns 'nvestigator_or_source', 'pdf', 'href_formula' and 'href'
df = df.drop(columns=['investigator_or_source', 'pdf', 'href_formula', 'href'])

# Check the result
df.head()

case_number         date    year        type    country             area  \
0    2020.08.20  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1    2020.08.14  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2    2020.08.10  10-Aug-2020  2020.0    Provoked        USA          Florida   
3    2020.08.02  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  2020.07.31.c  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity             name  \
0         New Smyrna Beach, Volusia County  Boogie boarding   Carolina Jones   
1             Shelly Beach, Port Macquarie          Surfing  Chantelle Doyle   
2  Off Gasparilla Island, Charlotte County          Fishing             male   
3                   Candle Reef, St. Croix       Snorkeling     Melony Klein   
4         New Smyrna Beach, Volusia County          Surfing      Megan Tossi   

  sex  age                                             injury fatal   time  \
0   F   50                      Minor lacerations to left leg     N  11h00   
1   F   35      Lacerations to right calf and posterior thigh     N  09h30   
2   M   55  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3   F  NaN                      Lacerations to hand and wrist     N  14h00   
4   F   22                               Lacerations to foot      N  17h00   

                species case_number.1 case_number.2  original_order  
0                   NaN    2020.08.20    2020.08.20          6559.0  
1  White shark, 2-to 3m    2020.08.14    2020.08.14          6558.0  
2    Blacktip shark, 6'    2020.08.10    2020.08.10          6557.0  
3       Nurse shark, 5'    2020.08.02    2020.08.02          6556.0  
4                   NaN  2020.07.31.c  2020.07.31.c          6555.0

The content of columns `case_number`, `case_number.1`, `case_number.2` and `original_order` is just a way to reference each case and is not relevant to the analysis. So, they will also be removed.

In [9]:
# Remove the columns 'case_number', 'case_number.1', 'case_number.2' and 'original_order'
df = df.drop(columns=['case_number', 'case_number.1', 'case_number.2', 'original_order'])

# Check the result
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity             name  \
0         New Smyrna Beach, Volusia County  Boogie boarding   Carolina Jones   
1             Shelly Beach, Port Macquarie          Surfing  Chantelle Doyle   
2  Off Gasparilla Island, Charlotte County          Fishing             male   
3                   Candle Reef, St. Croix       Snorkeling     Melony Klein   
4         New Smyrna Beach, Volusia County          Surfing      Megan Tossi   

  sex  age                                             injury fatal   time  \
0   F   50                      Minor lacerations to left leg     N  11h00   
1   F   35      Lacerations to right calf and posterior thigh     N  09h30   
2   M   55  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3   F  NaN                      Lacerations to hand and wrist     N  14h00   
4   F   22                               Lacerations to foot      N  17h00   

                species  
0                   NaN  
1  White shark, 2-to 3m  
2    Blacktip shark, 6'  
3       Nurse shark, 5'  
4                   NaN

After removing the unnecessary columns, the information about the new dataframe is shown below.

The column `name` is also irrelevant, since there will not be a specific analysis of each individual. So, this column can be removed.

In [10]:
# Remove column 'name'
df = df.drop(columns='name')

# Check the result
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity sex  age  \
0         New Smyrna Beach, Volusia County  Boogie boarding   F   50   
1             Shelly Beach, Port Macquarie          Surfing   F   35   
2  Off Gasparilla Island, Charlotte County          Fishing   M   55   
3                   Candle Reef, St. Croix       Snorkeling   F  NaN   
4         New Smyrna Beach, Volusia County          Surfing   F   22   

                                              injury fatal   time  \
0                      Minor lacerations to left leg     N  11h00   
1      Lacerations to right calf and posterior thigh     N  09h30   
2  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3                      Lacerations to hand and wrist     N  14h00   
4                               Lacerations to foot      N  17h00   

                species  
0                   NaN  
1  White shark, 2-to 3m  
2    Blacktip shark, 6'  
3       Nurse shark, 5'  
4                   NaN

In [11]:
# Check information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25792 entries, 0 to 25791
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6558 non-null   object 
 1   year      6556 non-null   float64
 2   type      6552 non-null   object 
 3   country   6508 non-null   object 
 4   area      6091 non-null   object 
 5   location  6008 non-null   object 
 6   activity  6002 non-null   object 
 7   sex       5987 non-null   object 
 8   age       3660 non-null   object 
 9   injury    6528 non-null   object 
 10  fatal     6006 non-null   object 
 11  time      3139 non-null   object 
 12  species   3610 non-null   object 
dtypes: float64(1), object(12)
memory usage: 2.6+ MB


## Rows

### Rows containing only missing values

There is total of 25792 lines, but all columns do not have even 9,000 rows with `non-NaN`. So,as mentioned before, there are a lot of rows that only contains missing values `NaN` and they can be removed. 

In [12]:
# Remove all rows that contains only NaN
df = df.dropna(how='all')

# Check the result
print(f'Number of rows after removing the rows containing only missing values: {df.shape[0]}\n')

# Check information about the dataframe
df.info()

Number of rows after removing the rows containing only missing values: 6558

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6558 entries, 0 to 6557
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6558 non-null   object 
 1   year      6556 non-null   float64
 2   type      6552 non-null   object 
 3   country   6508 non-null   object 
 4   area      6091 non-null   object 
 5   location  6008 non-null   object 
 6   activity  6002 non-null   object 
 7   sex       5987 non-null   object 
 8   age       3660 non-null   object 
 9   injury    6528 non-null   object 
 10  fatal     6006 non-null   object 
 11  time      3139 non-null   object 
 12  species   3610 non-null   object 
dtypes: float64(1), object(12)
memory usage: 717.3+ KB


### Duplicate rows

In [13]:
# Check if there are duplicate rows
df.duplicated().sum()

7

In [14]:
# Check duplicated rows
df[df.duplicated(keep=False)]

date    year        type         country  \
3577  Reported 26-Jun-1972  1972.0  Unprovoked       AUSTRALIA   
3578  Reported 26-Jun-1972  1972.0  Unprovoked       AUSTRALIA   
3622                  1971  1971.0  Unprovoked            IRAN   
3623                  1971  1971.0  Unprovoked            IRAN   
4490              Aug-1956  1956.0    Provoked  UNITED KINGDOM   
4491              Aug-1956  1956.0    Provoked  UNITED KINGDOM   
4940             Fall 1943  1943.0  Unprovoked             USA   
4941             Fall 1943  1943.0  Unprovoked             USA   
5719  Reported 10-Oct-1906  1906.0  Unprovoked             USA   
5721  Reported 10-Oct-1906  1906.0  Unprovoked             USA   
5962                  1890  1890.0  Unprovoked           INDIA   
5963                  1890  1890.0  Unprovoked           INDIA   
6215                  1863  1863.0  Unprovoked          CEYLON   
6216                  1863  1863.0  Unprovoked          CEYLON   

                    area                                      location  \
3577          Queensland                                 Pancake Creek   
3578          Queensland                                 Pancake Creek   
3622  Khuzestan Province                     Ahvaz, on the Karun River   
3623  Khuzestan Province                     Ahvaz, on the Karun River   
4490            Cornwall                                    The Lizard   
4491            Cornwall                                    The Lizard   
4940              Hawaii  Midway Island, Northwestern Hawaiian Islands   
4941              Hawaii  Midway Island, Northwestern Hawaiian Islands   
5719              Hawaii                                           NaN   
5721              Hawaii                                           NaN   
5962          Tamil Nadu                                     Tuticorin   
5963          Tamil Nadu                                     Tuticorin   
6215                 NaN                                           NaN   
6216                 NaN                                           NaN   

                                        activity sex  age  \
3577                                         NaN   M  NaN   
3578                                         NaN   M  NaN   
3622                                         NaN   M  NaN   
3623                                         NaN   M  NaN   
4490  Attempting to kill a shark with explosives   M  NaN   
4491  Attempting to kill a shark with explosives   M  NaN   
4940                                Spearfishing   M  NaN   
4941                                Spearfishing   M  NaN   
5719                                    Swimming   M  NaN   
5721                                    Swimming   M  NaN   
5962                                      Diving   M  NaN   
5963                                      Diving   M  NaN   
6215                              Fell overboard   M  NaN   
6216                              Fell overboard   M  NaN   

                        injury    fatal time         species  
3577                     FATAL        Y  NaN             NaN  
3578                     FATAL        Y  NaN             NaN  
3622                  Survived        N  NaN             NaN  
3623                  Survived        N  NaN             NaN  
4490  FATAL, PROVOKED INCIDENT        Y  NaN             NaN  
4491  FATAL, PROVOKED INCIDENT        Y  NaN             NaN  
4940  Calf nipped in each case        N  NaN  "small sharks"  
4941  Calf nipped in each case        N  NaN  "small sharks"  
5719                     FATAL        Y  NaN             NaN  
5721                     FATAL        Y  NaN             NaN  
5962                No details  UNKNOWN  NaN             NaN  
5963                No details  UNKNOWN  NaN             NaN  
6215                     FATAL        Y  NaN             NaN  
6216                     FATAL        Y  NaN             NaN

As shown above, there are some rows that are duplicated. Therefore, they can be removed.

In [15]:
# Remove duplicated rows
df = df.drop_duplicates(keep='first')

# Check the result
df.duplicated().sum()

0

In [16]:
# Check the new number of rows
n_rows = df.shape[0]
print(f'After removing the duplicates, the dataframe has {n_rows} rows.')

After removing the duplicates, the dataframe has 6551 rows.


# Specific Data Cleaning & Manipulation - Columns

## Gender

The gender is represented by the column `sex`.

In [17]:
# Check the unique values in the column
print(df.sex.unique())

['F' 'M' nan 'M ' 'lli' 'M x 2' 'N' '.']


In [18]:
# Check the number of each value
df.sex.value_counts()

M        5280
F         693
N           2
M           2
.           1
M x 2       1
lli         1
Name: sex, dtype: int64

In [19]:
# Remove unnecessary spaces in the values
df.sex = df.sex.str.strip()

# Check the result
df.sex.value_counts()

M        5282
F         693
N           2
.           1
M x 2       1
lli         1
Name: sex, dtype: int64

In [20]:
# Check the rows with values that are not 'M', nor 'F', nor NaN
print(f'They represent only {(len(df[(df.sex != "M") & (df.sex != "F") & (df.sex == df.sex)].index) / n_rows) * 100:.2f}',
      f'% of the dataset.', sep='')
df[(df.sex != 'M') & (df.sex != 'F') & (df.sex == df.sex)]

They represent only 0.08% of the dataset.


date    year          type           country  \
1867           11-Nov-2004  2004.0    Unprovoked               USA   
3205           23-Oct-1962  1982.0  Sea Disaster               USA   
5191           11-Jul-1934  1934.0    Watercraft         AUSTRALIA   
5690  Reported 02-Jun-1908  1908.0  Sea Disaster  PAPUA NEW GUINEA   
6386  Reported 18-Dec-1801  1801.0      Provoked               NaN   

                 area                                        location  \
1867       California  Bunkers, Humboldt Bay, Eureka, Humboldt County   
3205   Carolina coast                                             NaN   
5191  New South Wales                                        Cronulla   
5690      New Britain                                          Matupi   
6386              NaN                                             NaN   

                              activity    sex  age  \
1867                           Surfing    lli   38   
3205  Yacht Trashman capsized in storm  M x 2  NaN   
5191                           Fishing      N  NaN   
5690                                 .      .  NaN   
6386   Standing on landed shark's tail      N  NaN   

                                                 injury  fatal   time  \
1867                 Lacerations to hand, knee & thigh       N  13h30   
3205                                              FATAL  Y x 2    NaN   
5191  No injury to occupants Sharks continually foll...      N    NaN   
5690  Remains of 3 humans recovered from shark, but ...      Y    NaN   
6386                           FATAL, PROVOKED INCIDENT      Y    NaN   

                        species  
1867    5.5 m [18'] white shark  
3205                        NaN  
5191         Blue pointer, 11'   
5690  Allegedly a 33-foot shark  
6386                  12' shark

Given the low percentage, they will not be considered in the analysis, in other other, they will be removed.

In [21]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(df[(df.sex != 'M') & (df.sex != 'F') & (df.sex == df.sex)].index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6551 rows. After the removal, there is 6546 rows.


In [22]:
# Check rows with missing values in the 'sex' column
df[df.sex.isna()]

date    year          type           country  \
8              29-Jul-2020  2020.0    Watercraft         AUSTRALIA   
22             20-Jun-2020  2020.0    Unprovoked           BAHAMAS   
32             17 May 2020  2020.0    Unprovoked  FRENCH POLYNESIA   
48             05-Feb-2020  2020.0    Unprovoked               USA   
65             20-Dec-2019  2019.0      Provoked         AUSTRALIA   
...                    ...     ...           ...               ...   
6468           Before 2004     0.0    Watercraft        MOZAMBIQUE   
6493  No date, Before 1963     0.0    Unprovoked         SINGAPORE   
6515             1941-1945     0.0  Sea Disaster               NaN   
6534   Between 1918 & 1939     0.0    Unprovoked           REUNION   
6543           Before 1917     0.0    Unprovoked              FIJI   

                    area                                           location  \
8               Tasmania                                       Tenth Island   
22                Exumas                              Pig Beach, Pig Island   
32                Tahiti                                     Vallée Blanche   
48                  Maui                                                NaN   
65       New South Wales                                       Shellharbour   
...                  ...                                                ...   
6468  Inhambane Province                                      Off Inhambane   
6493                 NaN  Keppel Harbor,  2 miles from Singapore city ce...   
6515                 NaN                                                NaN   
6534         Saint-Denis                                          Barachois   
6543        Moala Island                                                NaN   

                                        activity  sex  age  \
8                                    Sightseeing  NaN  NaN   
22                                           NaN  NaN  NaN   
32                                    Snorkeling  NaN  NaN   
48                      Stand-Up Paddle boarding  NaN  NaN   
65                                       Fishing  NaN  NaN   
...                                          ...  ...  ...   
6468                                     Fishing  NaN  NaN   
6493                                    Swimming  NaN  NaN   
6515  A group of survivors on a raft for 17-days  NaN  NaN   
6534                                    Swimming  NaN  NaN   
6543        Wreck of  large double sailing canoe  NaN  NaN   

                                                 injury fatal            time  \
8     No injury to occupants, injury to shark attemp...   NaN           09h08   
22                                                  NaN     N             NaN   
32                                           Leg bitten     N             NaN   
48                    No injury, but paddleboard bitten     N           09h40   
65                                    PROVOKED INCIDENT   NaN             NaN   
...                                                 ...   ...             ...   
6468          No injury to occupants, shark bumped boat     N             NaN   
6493                                          Recovered     N             NaN   
6515  FATAL, shark leapt into raft and bit the man w...     Y  Late afternoon   
6534                                              FATAL     Y             NaN   
6543  FATAL, 18 people  were killed by sharks, 2 sur...     Y             NaN   

               species  
8      White shark, 4m  
22                 NaN  
32      Blacktip shark  
48         Tiger shark  
65         White shark  
...                ...  
6468       Whale shark  
6493               NaN  
6515  1.2 m [4'] shark  
6534               NaN  
6543               NaN  

[571 rows x 13 columns]

In [23]:
# Check % of missing values
df.sex.isna().mean()*100

8.722884204094102

For now, they will be remain in the dataframe, but they may be removed later.

## Age

### Data Cleaning

In [24]:
# Check the unique values in the column
print(f'There are {len(list(df.age.unique()))} unique values in this column.\n')
print(df.age.unique())

There are 237 unique values in this column.

[50 35 55 nan 22 14 28 38 4 63 23 11 12 10 29 15 36 7 16 30 60 18 9 26 57
 'Teen' 24 59 13 75 21 '30s' 45 33 17 37 70 44 '28 & 22' 32 20 51
 '22, 57, 31' '60s' 40 49 "20's" 43 8 64 19 65 67 53 34 25 58 74 46 41 31
 '9 & 60' 48 '20s' 42 39 56 61 'a minor' 6 62 52 54 69 '40s' 3 82 73 68 47
 66 72 27 71 '38' '39' '23' '32' '52' '68' '12' '18' '19' '43' '47' '6'
 '37' '9' '36' '10' '16' '13' '11' '17' '14' '30' '50' '29' '65' '63' '26'
 '71' '48' '70' '58' '18 months' '22' '41' '35' '57' '20' '24' '34' '15'
 '44' '53' '7' '40' '28' '33' '31' '45' '50s' '8' '51' '61' '42' '25'
 'teen' '66' '21' '77' '46' '60' '74' '55' '27' '3' '56' '64' '28 & 26'
 '62' '5' '49' '54' '86' '59' '18 or 20' '12 or 13' '46 & 34'
 '28, 23 & 30' 'Teens' 77 '36 & 26' '8 or 10' 84 '\xa0 ' ' ' '30 or 36'
 '6½' '21 & ?' '33 or 37' 'mid-30s' '23 & 20' 5 ' 30' '7      &    31'
 ' 28' '20?' "60's" '69' '32 & 30' '16 to 18' '87' '67' 'Elderly'
 'mid-20s' 'Ca. 33' '74 ' '45 ' '

In [25]:
# Check the number of each value
df.age.value_counts()

19             86
16             86
18             82
17             82
17             80
               ..
a minor         1
18 or 20        1
?    &   14     1
X               1
1               1
Name: age, Length: 236, dtype: int64

In [26]:
# Convert all values in this columnn to string if it is not a missing value
df.age = df.age.apply(lambda x : str(x) if x == x else x)

# Check result
df.age.unique()

array(['50', '35', '55', nan, '22', '14', '28', '38', '4', '63', '23',
       '11', '12', '10', '29', '15', '36', '7', '16', '30', '60', '18',
       '9', '26', '57', 'Teen', '24', '59', '13', '75', '21', '30s', '45',
       '33', '17', '37', '70', '44', '28 & 22', '32', '20', '51',
       '22, 57, 31', '60s', '40', '49', "20's", '43', '8', '64', '19',
       '65', '67', '53', '34', '25', '58', '74', '46', '41', '31',
       '9 & 60', '48', '20s', '42', '39', '56', '61', 'a minor', '6',
       '62', '52', '54', '69', '40s', '3', '82', '73', '68', '47', '66',
       '72', '27', '71', '18 months', '50s', 'teen', '77', '28 & 26', '5',
       '86', '18 or 20', '12 or 13', '46 & 34', '28, 23 & 30', 'Teens',
       '36 & 26', '8 or 10', '84', '\xa0 ', ' ', '30 or 36', '6½',
       '21 & ?', '33 or 37', 'mid-30s', '23 & 20', ' 30',
       '7      &    31', ' 28', '20?', "60's", '32 & 30', '16 to 18',
       '87', 'Elderly', 'mid-20s', 'Ca. 33', '74 ', '45 ', '21 or 26',
       '20 ', '>50', '

Converting all the values to the same data type make it easier to manipulate them.

In [27]:
len(df.age.unique())

164

In [28]:
# Check rows that contains '&'
df[df.age.str.contains('&', na=False)]

date    year          type             country  \
80             29-Oct-2019  2019.0    Unprovoked           AUSTRALIA   
168            10-Jan-2019  2019.0       Invalid           AUSTRALIA   
940            10-Mar-2013  2013.0    Unprovoked         PHILIPPINES   
1749  Reported 28-Jan-2006  2006.0    Watercraft      ATLANTIC OCEAN   
1751           23-Jan-2006  2006.0    Watercraft      ATLANTIC OCEAN   
1986           14-Sep-2003  2003.0    Watercraft        SOUTH AFRICA   
2168           12-Aug-2001  2001.0    Unprovoked            THAILAND   
2269           10-Aug-2000  2000.0       Invalid                 USA   
2665  Reported 11-Sep-1994  1994.0  Sea Disaster                 USA   
2754           04-Jan-1993  1993.0    Watercraft       CARIBBEAN SEA   
3709           24-Aug-1968  1968.0    Unprovoked                 USA   
3767           10-Aug-1967  1967.0       Invalid           AUSTRALIA   
3992              Aug-1963  1963.0    Unprovoked  DOMINICAN REPUBLIC   
4130  Reported 06-Sep-1961  1961.0      Provoked               ITALY   
4236           07-Jun-1960  1960.0       Invalid                 USA   
4248           14-Apr-1960  1960.0       Invalid             BERMUDA   
4259           29-Feb-1960  1960.0    Unprovoked           AUSTRALIA   
4285           28-Dec-1959  1959.0    Watercraft           AUSTRALIA   
4289           11-Dec-1959  1959.0      Provoked           AUSTRALIA   
4328                 21764  1959.0       Invalid               ITALY   
4368           02-Jan-1959  1959.0    Unprovoked          MOZAMBIQUE   
4840           14-Jan-1947  1947.0    Watercraft           AUSTRALIA   
5252           06-Jan-1932  1932.0      Provoked              MEXICO   
5434           04-Jul-1924  1924.0    Watercraft                 USA   

                        area  \
80                Queensland   
168               Queensland   
940                  Palawan   
1749  300 miles from Antigua   
1751     800 miles from land   
1986   Western Cape Province   
2168         Rayong Province   
2269                 Florida   
2665                 Florida   
2754                     NaN   
3709                 Florida   
3767         New South Wales   
3992           Santo Domingo   
4130         Venice Province   
4236              California   
4248                33N, 68W   
4259                Tasmania   
4285         New South Wales   
4289                Victoria   
4328                 Tuscany   
4368         Maputo Province   
4840         New South Wales   
5252         Baja California   
5434              California   

                                               location  \
80                        Off Airlie Beach, Whitsundays   
168         Catseye Beach, Hamilton Island, Whitsundays   
940                                    Off Likas Island   
1749                                                NaN   
1751                                                NaN   
1986                                      Melkbosstrand   
2168                                 Laem Mae Pim Beach   
2269                                       Florida Keys   
2665                                    Florida Straits   
2754                             Off Dominican Republic   
3709                 Half mile north of Juno Beach Pier   
3767                                         Jervis Bay   
3992                                     Rio Haina Port   
4130                                           Chioggia   
4236   10 miles off Santa Barbara, Santa Barbara County   
4248  Royal Canadian Navy CS2F-1 aircraft ditched in...   
4259                                          Ralph Bay   
4285                                  Leichardt, Sydney   
4289                                  Altona, Melbourne   
4328                   Cala del Corvo, Isola del Giglio   
4368                                  Off Inhaca Island   
4840                        Berry's Bay, Sydney Harbour   
5252                                      Descano Point   
5434

In [29]:
# Number of rows that contain "&"
r_clean = len(df[df.age.str.contains("&", na=False)])
print(f'There are {r_clean} rows containing "&" and they represent only {(r_clean / n_rows) * 100:.2f}% of the', end=' '
      f'dataset.\nSince they have ages of more than one person, they will be removed from the dataframe', sep=' ')

There are 24 rows containing "&" and they represent only 0.37% of the dataset.
Since they have ages of more than one person, they will be removed from the dataframe

In [30]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(df[df.age.str.contains('&', na=False)].index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6546 rows. After the removal, there is 6522 rows.


In [31]:
# Check the unique values
print(f'Now, there are {len(list(df.age.unique()))} unique values in this column.\n')
print(df.age.unique())

Now, there are 140 unique values in this column.

['50' '35' '55' nan '22' '14' '28' '38' '4' '63' '23' '11' '12' '10' '29'
 '15' '36' '7' '16' '30' '60' '18' '9' '26' '57' 'Teen' '24' '59' '13'
 '75' '21' '30s' '45' '33' '17' '37' '70' '44' '32' '20' '51' '22, 57, 31'
 '60s' '40' '49' "20's" '43' '8' '64' '19' '65' '67' '53' '34' '25' '58'
 '74' '46' '41' '31' '48' '20s' '42' '39' '56' '61' 'a minor' '6' '62'
 '52' '54' '69' '40s' '3' '82' '73' '68' '47' '66' '72' '27' '71'
 '18 months' '50s' 'teen' '77' '5' '86' '18 or 20' '12 or 13' 'Teens'
 '8 or 10' '84' '\xa0 ' ' ' '30 or 36' '6½' '33 or 37' 'mid-30s' ' 30'
 ' 28' '20?' "60's" '16 to 18' '87' 'Elderly' 'mid-20s' 'Ca. 33' '74 '
 '45 ' '21 or 26' '20 ' '>50' '18 to 22' 'adult' '9 months' '25 to 35' '1'
 '(adult)' '25 or 28' 'X' '"middle-age"' '13 or 18' '2 to 3 months'
 'MAKE LINE GREEN' ' 43' '81' '"young"' '7 or 8' '78' 'F' 'Both 11'
 '9 or 10' 'young' '  ' 'A.M.' '10 or 12' '31 or 33' '2½' '13 or 14']


In [32]:
# Check rows that the age values does not make sense or can not be categorized
df[df.age.apply(lambda x : True if x in ['\xa0 ', 'X', ' ', '20?', '>50', '13 or 18', 'MAKE LINE GREEN', 'Ca. 33',
                                         '"young"', 'F', 'Both 11', 'young', '  ', 'A.M.'] else False)]

date    year          type           country  \
2067  27-Sep-2002  2002.0      Provoked               USA   
2068  27-Sep-2002  2002.0      Provoked             TONGA   
2684  06-Apr-1994  1994.0    Unprovoked           BAHAMAS   
3229     May 1982  1982.0       Invalid               USA   
3355  29-Dec-1978  1978.0    Unprovoked         AUSTRALIA   
3612  11-Apr-1971  1971.0    Unprovoked      SOUTH AFRICA   
4302  07-Oct-1959  1959.0    Unprovoked        MOZAMBIQUE   
4325  11-Aug-1959  1959.0    Unprovoked             JAPAN   
4643  08-Jan-1953  1953.0    Watercraft         AUSTRALIA   
4700  24-Mar-1951  1951.0    Unprovoked         AUSTRALIA   
4929  14-May-1943  1943.0  Sea Disaster         AUSTRALIA   
5054  12-Jan-1939  1939.0    Unprovoked         AUSTRALIA   
5139  23-Feb-1936  1936.0      Provoked         AUSTRALIA   
5337  27-Jan-1929  1929.0    Unprovoked         AUSTRALIA   
5423  10-Mar-1925  1925.0    Unprovoked  PAPUA NEW GUINEA   
6398         1779  1779.0    Unprovoked               USA   

                       area                                  location  \
2067                Florida                  Key Largo, Monroe County   
2068                 Vava'u             Swimming with humpback whales   
2684                    NaN                                   Rum Cay   
3229         South Carolina  Daws Island, Broad River (near Beaufort)   
3355             Queensland                             Bribie Island   
3612  Western Cape Province                               Buffels Bay   
4302        Maputo Province              Xefina Island, Bay of Maputo   
4325    Wakayama Prefecture            Isonoura Beach, Wakayama City    
4643               Tasmania                                   Wynyard   
4700        New South Wales                                    Sydney   
4929             Queensland                              Off Brisbane   
5054        New South Wales                            Clarence River   
5139        New South Wales                      Angourie, near Yamba   
5337             Queensland     Alma Bay, Magnetic Island, Townsville   
5423       Central Province                                      Hula   
6398                 Hawaii                            Maliu, Hawai'i   

                                               activity  sex              age  \
2067                                            Fishing    M                    
2068                                           Swimming    M                    
2684                                             Wading    M              20?   
3229                                                NaN    F           Ca. 33   
3355                                                NaN    M                    
3612                                           Swimming    M              >50   
4302                                           Swimming    M                X   
4325                                           Swimming    M         13 or 18   
4643                                            Fishing  NaN  MAKE LINE GREEN   
4700                       Fishing, casting in the surf    M          "young"   
4929  Hospital Ship Centaur torpedoed & sunk by the ...    M                F   
5054                                    Scooping prawns    M          Both 11   
5139      Touching the mouth of a supposedly dead shark    F            young   
5337                                           Swimming    M                    
5423                                           Swimming    M             A.M.   
6398                                            Surfing    M            young   

                                                 injury fatal         time  \
2067             Left thumb lacerated PROVOKED INCIDENT     N    Afternoon   
2068                 Thigh lacerated  PROVOKED INCIDENT     N        15h00   
2684                            Lower right  leg bitten     N        10h30   
3229           Human remains recovered from

In [33]:
# Number of rows that contain "&"
r_clean = len(df[df.age.apply(lambda x : True if x in ['\xa0 ', 'X', ' ', '20?', '>50', '13 or 18', 'MAKE LINE GREEN',
                                                       'Ca. 33', '"young"', 'F', 'Both 11', 'young', '  ', 'A.M.'] 
                              else False)].index)
print(f'There are {r_clean} rows containing "&" and they represent only {(r_clean / n_rows) * 100:.2f}% of the', end=' '
      f'dataset.\nSince they can not be categorized, they will be removed from the dataframe', sep=' ')

There are 16 rows containing "&" and they represent only 0.25% of the dataset.
Since they can not be categorized, they will be removed from the dataframe

In [34]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(df[df.age.apply(lambda x : True if x in ['\xa0 ', 'X', ' ', '20?', '>50', '13 or 18', 'MAKE LINE GREEN', 
                                                      'Ca. 33', '"young"', 'F', 'Both 11', 'young', '  ', 'A.M.', '30s',
                                                     'a minor'] 
                             else False)].index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6522 rows. After the removal, there is 6498 rows.


In [35]:
# Check the unique values
print(f'Now, there are {len(list(df.age.unique()))} unique values in this column.\n')
print(df.age.unique())

Now, there are 125 unique values in this column.

['50' '35' '55' nan '22' '14' '28' '38' '4' '63' '23' '11' '12' '10' '29'
 '15' '36' '7' '16' '30' '60' '18' '9' '26' '57' 'Teen' '24' '59' '13'
 '75' '21' '45' '33' '17' '37' '70' '44' '32' '20' '51' '22, 57, 31' '60s'
 '40' '49' "20's" '43' '8' '64' '19' '65' '67' '53' '34' '25' '58' '74'
 '46' '41' '31' '48' '20s' '42' '39' '56' '61' 'a minor' '6' '62' '52'
 '54' '69' '40s' '3' '82' '73' '68' '47' '66' '72' '27' '71' '18 months'
 '50s' 'teen' '77' '5' '86' '18 or 20' '12 or 13' 'Teens' '8 or 10' '84'
 '30 or 36' '6½' '33 or 37' 'mid-30s' ' 30' ' 28' "60's" '16 to 18' '87'
 'Elderly' 'mid-20s' '74 ' '45 ' '21 or 26' '20 ' '18 to 22' 'adult'
 '9 months' '25 to 35' '1' '(adult)' '25 or 28' '"middle-age"'
 '2 to 3 months' ' 43' '81' '7 or 8' '78' '9 or 10' '10 or 12' '31 or 33'
 '2½' '13 or 14']


In [36]:
# Check % of missing values
df.age.isna().mean()*100

44.429055093875036

The percentage of missing values is too high and, if they are removed, it would greatly impact the dataset. So, for now, they will remain in the dataframe and they will be analyzed again later.

### Categorizing the age

The column age will be categorized into:

<table>
    <thead>
        <tr>
            <th>AGE</th>
            <th>CATEGORY</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>< 11</td>
            <td>Child</td>
        </tr>
        <tr>
            <td>12 - 17</td>
            <td>Teenager</td>
        </tr>
        <tr>
            <td>18 - 35</td>
            <td>Young Adult</td>
        </tr>
        <tr>
            <td>36 - 65</td>
            <td>Adult</td>
        </tr>
        <tr>
            <td>> 65</td>
            <td>Elder</td>
        </tr>
        <tr>
            <td>NaN</td>
            <td>No category</td>
        </tr>
    </tbody>
</table>

In [48]:
# Create a function to categorize the ages
def cat_ages(age):
    '''
    
    '''
    
    try:
        age = int(age)
        if age > 65:
            return 'Elder'
        elif age > 35:
            return 'Adult'
        elif age > 17:
            return 'Young Adult'
        elif age > 11:
            return 'Teenager'
        else:
            return 'Child'
    except:
        if age != age:
            return 'No category'
        elif age in ['18 months', '8 or 10', '6½', '9 months', '2 to 3 months', '7 or 8', '9 or 10', '10 or 12', '2½']:
            return 'Child'
        elif age in ['Teen', 'teen', '12 or 13', 'Teens', '16 to 18', '13 or 14']:
            return 'Teenager'
        elif age in ["20's", '20s', '18 or 20', 'mid-20s', '30 or 36', '21 or 26', '18 to 22', '25 to 35', '25 or 28',
                     '31 or 33']:
            return 'Young Adult'
        elif age in ['40s', '50s', 'mid-30s', '33 or 37', 'adult', '(adult)', '"middle-age"']:
            return 'Adult'
        elif age in ['60s', "60's", 'Elderly']:
            return 'Elder'

In [49]:
# Categorize the ages
df['cat_age'] = df.age.apply(cat_ages)

# Check the result
df.cat_age.value_counts(dropna=False)

No category    2887
Young Adult    1723
Adult           840
Teenager        751
Child           235
Elder            60
NaN               2
Name: cat_age, dtype: int64

In [51]:
# Check number of rows
df.cat_age.value_counts(dropna=False).sum() == n_rows

True

In [52]:
# Check information abou the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6498 entries, 0 to 6557
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6498 non-null   object 
 1   year      6496 non-null   float64
 2   type      6492 non-null   object 
 3   country   6449 non-null   object 
 4   area      6035 non-null   object 
 5   location  5954 non-null   object 
 6   activity  5946 non-null   object 
 7   sex       5932 non-null   object 
 8   age       3611 non-null   object 
 9   injury    6468 non-null   object 
 10  fatal     5952 non-null   object 
 11  time      3116 non-null   object 
 12  species   3571 non-null   object 
 13  cat_age   6496 non-null   object 
dtypes: float64(1), object(13)
memory usage: 761.5+ KB


In [53]:
# Check the dataset
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity sex  age  \
0         New Smyrna Beach, Volusia County  Boogie boarding   F   50   
1             Shelly Beach, Port Macquarie          Surfing   F   35   
2  Off Gasparilla Island, Charlotte County          Fishing   M   55   
3                   Candle Reef, St. Croix       Snorkeling   F  NaN   
4         New Smyrna Beach, Volusia County          Surfing   F   22   

                                              injury fatal   time  \
0                      Minor lacerations to left leg     N  11h00   
1      Lacerations to right calf and posterior thigh     N  09h30   
2  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3                      Lacerations to hand and wrist     N  14h00   
4                               Lacerations to foot      N  17h00   

                species      cat_age  
0                   NaN        Adult  
1  White shark, 2-to 3m  Young Adult  
2    Blacktip shark, 6'        Adult  
3       Nurse shark, 5'  No category  
4                   NaN  Young Adult

## Activity

## Extra - Year

In [43]:
# Check the unique values in the column
df.year.unique()

array([2020., 2019., 2018., 2017.,   nan, 2016., 2015., 2014., 2013.,
       2012., 2011., 2010., 2009., 2008., 2007., 2006., 2005., 2004.,
       2003., 2002., 2001., 2000., 1999., 1998., 1997., 1996., 1995.,
       1984., 1994., 1993., 1992., 1991., 1990., 1989., 1969., 1988.,
       1987., 1986., 1985., 1983., 1982., 1981., 1980., 1979., 1978.,
       1977., 1976., 1975., 1974., 1973., 1972., 1971., 1970., 1968.,
       1967., 1966., 1965., 1964., 1963., 1962., 1961., 1960., 1959.,
       1958., 1957., 1956., 1955., 1954., 1953., 1952., 1951., 1950.,
       1949., 1948., 1848., 1947., 1946., 1945., 1944., 1943., 1942.,
       1941., 1940., 1939., 1938., 1937., 1936., 1935., 1934., 1933.,
       1932., 1931., 1930., 1929., 1928., 1927., 1926., 1925., 1924.,
       1923., 1922., 1921., 1920., 1919., 1918., 1917., 1916., 1915.,
       1914., 1913., 1912., 1911., 1910., 1909., 1908., 1907., 1906.,
       1905., 1904., 1903., 1902., 1901., 1900., 1899., 1898., 1897.,
       1896., 1895.,

# Analyzing the dataset

## Gender

## Activity

## Age

## Age

In [ ]:
df.Age.unique()

In [ ]:
# Function to convert string to integer
def convert_str_to_int(age):
    try:
        age = int(age)
    except:
        age = -1
    return age

In [ ]:
# Number 
df_total = df.Date.count()
print(df_total)

# Convert the strings in column 'Age' to integers
df['age_int'] = df.Age.apply(convert_str_to_int)

# How many were not digits
print(df.loc[df.age_int == -1, :].age_int.value_counts() / 6558)
print(df.loc[df.age_int == -1, :].age_int.value_counts() / df_total)

# Check the result
df

In [ ]:
# Classification by age
df['age_cat'] = np.where(df['age_int'] > 65, 'Elder',
                        np.where(df['age_int'] > 35, 'Adult',
                                np.where(df['age_int'] > 17, 'Young Adult',
                                        np.where(df['age_int'] > 12, 'Teenager',
                                                 np.where(df['age_int'] == -1, '-', 'Child')))))

## Year

In [ ]:
df.Year.unique()

In [ ]:
# Convert the strings in column 'Year' to integers
df['year_int'] = df.Year.apply(convert_str_to_int)

# Check the result
df

In [ ]:
# Check % incidents occured before 1801
print(f'The incidents before 1801 represents only {(df[df.Year < 1801].Year.count() / df.Year.count()) * 100:.2f}%',
      f'of the dataset. So, only the years from 1800 will be analysed.', sep=' ')

In [ ]:
# Selecting only the years from 1800
df = df.loc[df['Year'] >= 1801, :]
df.info()

In [ ]:
df

In [ ]:
# Classification by year
df['century'] = np.where(df['year_int'] >= 2001, 21,
                            np.where(df['year_int'] >= 1901, 20, 19))

In [ ]:
df

## Gender

In [ ]:
# Check the column 'Sex'
df.Sex.unique()

In [ ]:
df.Sex.count()

In [ ]:
# % of each value
df.Sex.value_counts() / df.Sex.count()

In [ ]:
# % of each NaN
df.Sex.isna().sum()

In [ ]:
# Check number of missing values
print(f'The number of NaN in the column "Sex" is {df.Sex.isna().sum()}, which represents '
      f'{(df.Sex.isna().sum() / df.Sex.count())*100:.2f}% of the dataset.', sep=' ')

Since the NaN represents only a samll part of the dataset, the rows containing NaN in the column 'Sex' will be removed.

In [ ]:
# Remove rows that the column 'Sex' is a missing value
df[df.Sex.isna()]
df = df.drop(df[df.Sex.isna()].index)

In [ ]:
# Check infromation about the changed dataset
df.info()

In [ ]:
# Check values in the column 'Sex'
df.Sex.value_counts()

In [ ]:
# Remove unnecessary spaces in the values of the column 'Sex'
df['Sex'] = df['Sex'].str.strip()
df.Sex.value_counts()

In [ ]:
# Remove rows that the gender is 'N', 'lli', '.' or 'M x 2'
df = df.drop(df[(df.Sex == '.') | (df.Sex == 'N') | (df.Sex == 'lli') | ((df.Sex == 'M x 2'))].index)
df.Sex.value_counts()

In [ ]:
# Check infromation about the changed dataset
# 5 columns were removed
df.info()

In [ ]:
# Check column 'Sex' percentage
df_gener_total = df.Sex.value_counts().sum()
df.Sex.value_counts() / df_gener_total

In [ ]:
print(f'The males represent {(df.Sex.value_counts() / df_gener_total)[0]*100:.2f}% of the people who were'
      f'involved in incidents with sharks, while women represent only {(df.Sex.value_counts() / df_gener_total)[1]*100:.2f}%.',sep=' ')

In [ ]:
# Create a subset containing only females
df_male = df.loc[df['Sex'] == 'F']

In [ ]:
# Create a subset containing only males
df_male = df.loc[df['Sex'] == 'M']
df_male.info()

In [ ]:
# Check males age
df_male.age_cat.value_counts()

In [ ]:
df_male_total = df_male.Date.count()
df_male_total

In [ ]:
# Check males age
df_male.age_cat.value_counts() / df_male_total

In [ ]:
(df_male.age_cat.value_counts() / df_male_total)*100

In [ ]:
df_male_age = df_male.loc[df_male['age_cat'] != '-']
df_male_age_total = df_male_age.Date.count()
df_male_age.info()

In [ ]:
df_male_age_total

In [ ]:
df_male_age.age_cat.value_counts()

In [ ]:
df_male_age.age_cat.value_counts() / df_male_age_total

In [ ]:
(df_male_age.age_cat.value_counts() / df_male_age_total)*100

50% of male who were in a shark incident were Young Adults.

In [ ]:
df_male_ya = df_male_age[df_male_age['age_cat'] == 'Young Adult']
df_male_ya.info()

In [ ]:
df_male_ya_total = df_male_ya.Date.count()
df_male_ya.groupby(by='Activity').age_cat.count().sort_values(ascending=False)

In [ ]:
(df_male_ya.groupby(by='Activity').age_cat.count().sort_values(ascending=False) / df_male_ya_total)*100

**Young adult males surfing were the most involved in shark incidents.**

# Exporting dataset

In [ ]:
df

In [ ]:
#df.to_csv('sharks_clean.csv')

# Extra - Analysis though centuries

## Setup

In [ ]:
# Create subsets for each century
df_cen21 = df.loc[df['century'] == 21, :]
df_cen20 = df.loc[df['century'] == 20, :]
df_cen19 = df.loc[df['century'] == 19, :]

In [ ]:
df_cen21_total = df_cen21.Date.count()
print(df_cen21_total)
df_cen21.info()

In [ ]:
df_cen20_total = df_cen20.Date.count()
print(df_cen20_total)
df_cen20.info()

In [ ]:
df_cen19_total = df_cen19.Date.count()
print(df_cen19_total)
df_cen19.info()

We are not even 1/4 of the 21st century and the number of shark incidents getting closer to the number of incidents in the 20th century. However, we also have to consider that there may be a probability that not all incidents were registered in the last century. 

## Gender

In [ ]:
df_cen21.Sex.value_counts() / df_cen21_total

In [ ]:
(df_cen21.Sex.value_counts() / df_cen21_total) * 100

In [ ]:
df_cen20.Sex.value_counts() / df_cen20_total

In [ ]:
(df_cen20.Sex.value_counts() / df_cen20_total) * 100

In [ ]:
df_cen19.Sex.value_counts() / df_cen19_total

In [ ]:
(df_cen19.Sex.value_counts() / df_cen19_total) * 100

This result reflects each gender's behaviors in each century. Long time ago, women usually would just take care of the house and were not allowed to do many things.

## Activity

In [ ]:
df_cen21.Activity.value_counts() / df_cen21_total

In [ ]:
(df_cen21.Activity.value_counts() / df_cen21_total) * 100

In [ ]:
df_cen20.Activity.value_counts() / df_cen20_total

In [ ]:
(df_cen20.Activity.value_counts() / df_cen20_total) * 100

In [ ]:
df_cen19.Activity.value_counts() / df_cen19_total

In [ ]:
(df_cen19.Activity.value_counts() / df_cen19_total) * 100